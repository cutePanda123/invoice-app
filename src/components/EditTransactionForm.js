import React from 'react';
import PropTyes from 'prop-types';

class CreateTransactionForm extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            errorMessage: ''
        };
    }

    submitForm = (event) => {
        const {transaction, onFormSubmit} = this.props;
        const amount = this.amountInput.value.trim() * 1;
        const date = this.dateInput.value.trim();
        const title = this.titleInput.value.trim();
        const isEditMode = transaction.desc;

        if (amount && date && title) {
            if (amount < 0) {
                this.setState({
                    errorMessage: 'dollar amount cannot be negative'
                });
            } else {
                onFormSubmit({...transaction, amount, date, desc: title}, isEditMode)
            }
        } else {
            this.setState({
                errorMessage: 'input filed cannot be empty'
            });
        }
        event.preventDefault();
    }

    render() {
        return (
            <form onSubmit={(event) => {this.submitForm(event)}} noValidate>
                <div className="form-group">
                    <label htmlFor="title">Title</label>
                    <input
                        type="text"
                        className="form-control"
                        id="title"
                        aria-describedby="titleHelp"
                        placeholder="Enter title"
                        ref={(inputValue) => {
                            this.titleInput = inputValue;
                        }}
                        defaultValue={this.props.transaction.desc}
                    />
                    <small id="titleHelp" className="form-text text-muted">Title for your transaction.</small>
                </div>
                <div className="form-group">
                    <label htmlFor="amount">Dollar Amount</label>
                    <input
                        type="number"
                        className="form-control"
                        id="amount"
                        placeholder="Dollar Amount"
                        defaultValue={this.props.transaction.amount}
                        ref={(inputValue) => {
                            this.amountInput = inputValue;
                        }}
                    />
                    <small id="amountHelp" className="form-text text-muted">Dollar amount for your transaction.</small>
                </div>
                <div className="form-group">
                    <label htmlFor="date">Date</label>
                    <input
                        type="date"
                        className="form-control"
                        id="date"
                        placeholder="Date"
                        defaultValue={this.props.transaction.date}
                        ref={(inputValue) => {
                            this.dateInput = inputValue;
                        }}
                    />
                    <small id="dateHelp" className="form-text text-muted">Date for your transaction.</small>
                </div> 
                <button
                    type="submit"   
                    className="btn btn-primary"
                >Submit</button>
                <button
                    className="btn btn-secondary"
                    onClick={this.props.onFormCancel}
                >Cancel</button>
                {
                    this.state.errorMessage !== '' && 
                    <div className="alert alert-danger mt-5" role="alert">
                        {this.state.errorMessage}
                    </div>
                }
            </form>
        );
    }
}

CreateTransactionForm.propTypes = {
    transaction: PropTyes.object.isRequired,
    onFormSubmit: PropTyes.func.isRequired,
    onFormCancel: PropTyes.func.isRequired
}

CreateTransactionForm.defaultProps = {
    transaction: {}
}


export default CreateTransactionForm;