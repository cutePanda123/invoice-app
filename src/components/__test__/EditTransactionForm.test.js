import React from 'react';
import { mount } from 'enzyme';
import EditTransactionForm from '../EditTransactionForm';

const editModeProps = {
    transaction: {
        desc: "t",
        date: "2020-10-10",
        amount: 100
    },
    onFormSubmit: jest.fn(),
    onFormCancel: jest.fn()
};

const props = {
    onFormSubmit: jest.fn(),
    onFormCancel: jest.fn()
};

let wrapper = null;
let editModeWrapper = null;

describe('EditTransactionFrom test', () => {
    beforeEach(() => {
        wrapper = mount(<EditTransactionForm {...props} />);
        editModeWrapper = mount(<EditTransactionForm {...editModeProps} />);
    });

    it('snaptshot test', () => {
        expect(wrapper).toMatchSnapshot();
        expect(editModeWrapper).toMatchSnapshot();
    });

    it('should render three input fields', () => {
        const inputs = wrapper.find('.form-group');
        expect(inputs.length).toEqual(3);
        expect(getInputValue('#amount', wrapper)).toEqual('');
        expect(getInputValue('#date', wrapper)).toEqual('');
        expect(getInputValue('#desc', wrapper)).toEqual('');

        expect(getInputValue('#amount', editModeWrapper)).toEqual('100');
        expect(getInputValue('#date', editModeWrapper)).toEqual('2020-10-10');
        expect(getInputValue('#desc', editModeWrapper)).toEqual('t');
    });

    it('should render error message when input is empty', () => {
        const submitButton = wrapper.find('.btn-primary');
        wrapper.find('form').simulate('submit')
        expect(props.onFormSubmit).not.toHaveBeenCalled();
        expect(wrapper.state().errorMessage).not.toEqual('');
    });

    it('should render error message when input dollar amount is negative', () => {
        const submitButton = wrapper.find('.btn-primary');
        setInputValue('#date', wrapper, '2020-10-10');
        setInputValue('#desc', wrapper, 't');
        setInputValue('#amount', wrapper, '-10');
        wrapper.find('form').simulate('submit')
        expect(props.onFormSubmit).not.toHaveBeenCalled();
        expect(wrapper.state().errorMessage).not.toEqual('');
    });

    it('create mode: should trigger onFormSubmit if all input are valid', () => {
        setInputValue('#date', wrapper, '2020-10-10');
        setInputValue('#desc', wrapper, 't');
        setInputValue('#amount', wrapper, '10');
        wrapper.find('form').simulate('submit')
        expect(props.onFormSubmit).toHaveBeenCalledWith({
            date: '2020-10-10',
            desc: 't',
            amount: 10
        }, false);
        expect(wrapper.state().errorMessage).toEqual('');
    });

    it('edit mode: should trigger onFormSubmit if all input are valid', () => {
        setInputValue('#date', editModeWrapper, '2020-10-10');
        setInputValue('#desc', editModeWrapper, 'newTitle');
        setInputValue('#amount', editModeWrapper, '10');
        editModeWrapper.find('form').simulate('submit')
        expect(editModeProps.onFormSubmit).toHaveBeenCalledWith({
            date: '2020-10-10',
            desc: 'newTitle',
            amount: 10
        }, true);
        expect(editModeWrapper.state().errorMessage).toEqual('');
    });

    it('click cancel button should trigger onFormCancel', () => {
        editModeWrapper.find('.btn-secondary').simulate('click')
        expect(editModeProps.onFormCancel).toBeCalled();
    })
});

const getInputValue = (selector, domElement) => {
    return domElement.find(selector).instance().value;
};

const setInputValue = (selector, domElement, newValue) => {
    domElement.find(selector).instance().value = newValue;
};

