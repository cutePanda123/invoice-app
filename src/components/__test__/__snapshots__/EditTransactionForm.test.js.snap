// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditTransactionFrom test snaptshot test 1`] = `
<CreateTransactionForm
  onFormCancel={[MockFunction]}
  onFormSubmit={[MockFunction]}
  transaction={Object {}}
>
  <form
    noValidate={true}
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="title"
      >
        Title
      </label>
      <input
        aria-describedby="titleHelp"
        className="form-control"
        id="desc"
        placeholder="Enter title"
        type="text"
      />
      <small
        className="form-text text-muted"
        id="titleHelp"
      >
        Title for your transaction.
      </small>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="amount"
      >
        Dollar Amount
      </label>
      <input
        className="form-control"
        id="amount"
        placeholder="Dollar Amount"
        type="number"
      />
      <small
        className="form-text text-muted"
        id="amountHelp"
      >
        Dollar amount for your transaction.
      </small>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="date"
      >
        Date
      </label>
      <input
        className="form-control"
        id="date"
        placeholder="Date"
        type="date"
      />
      <small
        className="form-text text-muted"
        id="dateHelp"
      >
        Date for your transaction.
      </small>
    </div>
    <button
      className="btn btn-primary"
      type="submit"
    >
      Submit
    </button>
    <button
      className="btn btn-secondary"
      onClick={[MockFunction]}
    >
      Cancel
    </button>
  </form>
</CreateTransactionForm>
`;

exports[`EditTransactionFrom test snaptshot test 2`] = `
<CreateTransactionForm
  onFormCancel={[MockFunction]}
  onFormSubmit={[MockFunction]}
  transaction={
    Object {
      "amount": 100,
      "date": "2020-10-10",
      "desc": "t",
    }
  }
>
  <form
    noValidate={true}
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="title"
      >
        Title
      </label>
      <input
        aria-describedby="titleHelp"
        className="form-control"
        defaultValue="t"
        id="desc"
        placeholder="Enter title"
        type="text"
      />
      <small
        className="form-text text-muted"
        id="titleHelp"
      >
        Title for your transaction.
      </small>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="amount"
      >
        Dollar Amount
      </label>
      <input
        className="form-control"
        defaultValue={100}
        id="amount"
        placeholder="Dollar Amount"
        type="number"
      />
      <small
        className="form-text text-muted"
        id="amountHelp"
      >
        Dollar amount for your transaction.
      </small>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="date"
      >
        Date
      </label>
      <input
        className="form-control"
        defaultValue="2020-10-10"
        id="date"
        placeholder="Date"
        type="date"
      />
      <small
        className="form-text text-muted"
        id="dateHelp"
      >
        Date for your transaction.
      </small>
    </div>
    <button
      className="btn btn-primary"
      type="submit"
    >
      Submit
    </button>
    <button
      className="btn btn-secondary"
      onClick={[MockFunction]}
    >
      Cancel
    </button>
  </form>
</CreateTransactionForm>
`;
